examples/dg_advection.cpp

```diff
@@ -0,0 +1,988 @@
+            if (mpi.Root())
+            {
+                cout << "time step: " << ti << ", time: " << t << endl;
+            }
+
+            // 11. Extract the parallel grid function corresponding to the finite
+            //     element approximation U (the local solution on each processor).
+            *u = *U;
+
+            if (visualization)
+            {
+                sout << "parallel " << num_procs << " " << myid << "\n";
+                sout << "solution\n" << *pmesh << *u << flush;
+            }
+
+            if (visit)
+            {
+                dc->SetCycle(ti);
+                dc->SetTime(t);
+                dc->Save();
+            }
+
+            if (paraview)
+            {
+                pd->SetCycle(ti);
+                pd->SetTime(t);
+                pd->Save();
+            }
+
+#ifdef MFEM_USE_ADIOS2
+            // transient solutions can be visualized with ParaView
+            if (adios2)
+            {
+                adios2_dc->SetCycle(ti);
+                adios2_dc->SetTime(t);
+                adios2_dc->Save();
+            }
+#endif
+        }
+    }
+
+    // 12. Save the final solution in parallel. This output can be viewed later
+    //     using GLVis: "glvis -np <np> -m dg_advection-mesh -g dg_advection-final".
+    {
+        *u = *U;
+        ostringstream sol_name;
+        sol_name << "dg_advection-final." << setfill('0') << setw(6) << myid;
+        ofstream osol(sol_name.str().c_str());
+        osol.precision(precision);
+        u->Save(osol);
+    }
+
+    // 13. Calculate the DMD modes.
+    if (myid == 0 && rdim != -1 && ef != -1)
+    {
+        std::cout << "Both rdim and ef are set. ef will be ignored." << std::endl;
+    }
+
+    dmd_training_timer.Start();
+
+    if (rdim != -1)
+    {
+        if (myid == 0)
+        {
+            std::cout << "Creating DMD with rdim: " << rdim << std::endl;
+        }
+        dmd_U.train(rdim);
+    }
+    else if (ef != -1)
+    {
+        if (myid == 0)
+        {
+            std::cout << "Creating DMD with energy fraction: " << ef << std::endl;
+        }
+        dmd_U.train(ef);
+    }
+
+    dmd_training_timer.Stop();
+
+    dmd_prediction_timer.Start();
+
+    // 14. Predict the state at t_final using DMD.
+    if (myid == 0)
+    {
+        std::cout << "Predicting solution at t_final using DMD" << std::endl;
+    }
+    CAROM::Vector* result_u = dmd_U.predict(t_final, dt);
+
+    dmd_prediction_timer.Stop();
+
+    // 15. Calculate the relative error between the DMD final solution and the true solution.
+    Vector dmd_solution_u(result_u->getData(), result_u->dim());
+    Vector true_solution_u(U->GetData(), U->Size());
+    Vector diff_u(true_solution_u.Size());
+    subtract(dmd_solution_u, true_solution_u, diff_u);
+
+    double* diff_norm_u = new double[num_procs] {};
+    double proc_diff_norm_u = diff_u.Norml2();
+    MPI_Gather(&proc_diff_norm_u, 1, MPI_DOUBLE, diff_norm_u, 1, MPI_DOUBLE, 0, MPI_COMM_WORLD);
```

Comment:
Calculating global `l2` norms from local norms in this way is terribly inconvenient, but fortunately there is a function in MFEM to do this. You can use `sqrt(InnerProduct(MPI_COMM_WORLD, diff_u, diff_u));` to simplify all these examples.

Reply:
Thanks. I wish I knew this two weeks ago!